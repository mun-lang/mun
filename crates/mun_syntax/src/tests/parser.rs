use crate::SourceFile;

fn ok_snapshot_test(text: &str) {
    let text = text.trim().replace("\n    ", "\n");
    let file = SourceFile::parse(&text);
    let errors = file.errors();
    assert_eq!(
        &*errors,
        &[] as &[crate::SyntaxError],
        "There should be no errors\nAST:\n{}",
        file.debug_dump()
    );
    insta::assert_snapshot!(insta::_macro_support::AutoName, file.debug_dump(), &text);
}

#[test]
fn empty() {
    ok_snapshot_test(r#""#);
}

#[test]
fn function() {
    ok_snapshot_test(
        r#"
    // Source file comment

    // Comment that belongs to the function
    fn a() {}
    fn b(value:number) {}
    export fn c():never {}
    fn b(value:number):number {}"#,
    );
}

#[test]
fn block() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a;
        let b:i32;
        let c:string;
    }"#,
    );
}

#[test]
fn literals() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = true;
        let b = false;
        let c = 1;
        let d = 1.12;
        let e = "Hello, world!"
    }
    "#,
    );
}

#[test]
fn unary_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = --3;
        let b = !!true;
    }
    "#,
    )
}

#[test]
fn binary_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = 3+4*5
        let b = 3*4+10/2
    }
    "#,
    )
}

#[test]
fn expression_statement() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = "hello"
        let b = "world"
        let c
        b = "Hello, world!"
        !-5+2*(a+b);
        -3
    }
    "#,
    )
}

#[test]
fn function_calls() {
    ok_snapshot_test(
        r#"
    fn bar(i:number) { }
    fn foo(i:number) {
      bar(i+1)
    }
    "#,
    )
}

#[test]
fn patterns() {
    ok_snapshot_test(
        r#"
    fn main(_:number) {
       let a = 0;
       let _ = a;
    }
    "#,
    )
}

#[test]
fn compare_operands() {
    ok_snapshot_test(
        r#"
    fn main() {
        let _ = a==b;
        let _ = a == b;
        let _ = a != b;
        let _ = a < b;
        let _ = a > b;
        let _ = a <= b;
        let _ = a >= b;
    }
    "#,
    )
}

#[test]
fn if_expr() {
    ok_snapshot_test(
        r#"
    fn bar() {
        if true {};
        if true {} else {};
        if true {} else if false {} else {};
        if {true} {} else {}
    }
    "#,
    );
}

#[test]
fn block_expr() {
    ok_snapshot_test(
        r#"
    fn bar() {
        {3}
    }
    "#,
    );
}

#[test]
fn return_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        return;
        return 50;
    }
    "#,
    )
}

#[test]
fn loop_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        loop {}
    }"#,
    )
}
