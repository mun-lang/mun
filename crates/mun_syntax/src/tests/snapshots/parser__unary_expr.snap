---
source: crates/mun_syntax/src/tests/parser.rs
expression: "fn foo() {\n    let a = --3;\n    let b = !!true;\n}"
---
SOURCE_FILE@[0; 49)
  FUNCTION_DEF@[0; 49)
    FN_KW@[0; 2) "fn"
    WHITESPACE@[2; 3) " "
    NAME@[3; 6)
      IDENT@[3; 6) "foo"
    PARAM_LIST@[6; 8)
      L_PAREN@[6; 7) "("
      R_PAREN@[7; 8) ")"
    WHITESPACE@[8; 9) " "
    BLOCK_EXPR@[9; 49)
      L_CURLY@[9; 10) "{"
      WHITESPACE@[10; 15) "\n    "
      LET_STMT@[15; 27)
        LET_KW@[15; 18) "let"
        WHITESPACE@[18; 19) " "
        BIND_PAT@[19; 20)
          NAME@[19; 20)
            IDENT@[19; 20) "a"
        WHITESPACE@[20; 21) " "
        EQ@[21; 22) "="
        WHITESPACE@[22; 23) " "
        PREFIX_EXPR@[23; 26)
          MINUS@[23; 24) "-"
          PREFIX_EXPR@[24; 26)
            MINUS@[24; 25) "-"
            LITERAL@[25; 26)
              INT_NUMBER@[25; 26) "3"
        SEMI@[26; 27) ";"
      WHITESPACE@[27; 32) "\n    "
      LET_STMT@[32; 47)
        LET_KW@[32; 35) "let"
        WHITESPACE@[35; 36) " "
        BIND_PAT@[36; 37)
          NAME@[36; 37)
            IDENT@[36; 37) "b"
        WHITESPACE@[37; 38) " "
        EQ@[38; 39) "="
        WHITESPACE@[39; 40) " "
        PREFIX_EXPR@[40; 46)
          EXCLAMATION@[40; 41) "!"
          PREFIX_EXPR@[41; 46)
            EXCLAMATION@[41; 42) "!"
            LITERAL@[42; 46)
              TRUE_KW@[42; 46) "true"
        SEMI@[46; 47) ";"
      WHITESPACE@[47; 48) "\n"
      R_CURLY@[48; 49) "}"

